### PR TITLE
Support `direct_io`

### PR DESCRIPTION
**Pull request by DUOLabs333**

### CODE REVIEW

File: src/opts.c [hunk 1]
```diff
@@ -291,6 +291,7 @@ static void print_help(const char *progname) {
 	"    -o relaxed_permissions Disable permissions checks, but only if\n"
 	"                           running neither as UID=0 or GID=0\n"
 	"    -o statfs_omit_ro      do not count blocks of ro-branches\n"
+	"    -o direct_io           Support direct io\n"
```
Comment (rpodgorny) [inline src/opts.c:294]: please change "support" to "enable"

File: src/opts.h [hunk 2]
```diff
@@ -45,7 +46,8 @@ enum {
 	KEY_NOINITGROUPS,
 	KEY_RELAXED_PERMISSIONS,
 	KEY_STATFS_OMIT_RO,
-	KEY_VERSION
+	KEY_VERSION,
```
Comment (rpodgorny) [inline src/opts.h:49]: please leave KEY_VERSION at the end. move KEY_DIRECTO_IO to between KEY_CHROOT and KEY_COW (just as you did in your "switch" code).

File: src/unionfs.c [hunk 3]
```diff
@@ -36,6 +36,7 @@ static struct fuse_opt unionfs_opts[] = {
 	FUSE_OPT_KEY("statfs_omit_ro", KEY_STATFS_OMIT_RO),
 	FUSE_OPT_KEY("--version", KEY_VERSION),
 	FUSE_OPT_KEY("-V", KEY_VERSION),
+	FUSE_OPT_KEY("direct_io",KEY_DIRECT_IO),
```
Comment (rpodgorny) [inline src/unionfs.c:39]: ...same as above (leave "version" as the last thing).

File: src/fuse_ops.c [hunk 4]
```diff
@@ -399,6 +403,7 @@ static int unionfs_open(const char *path, struct fuse_file_info *fi) {
 	DBG("%s\n", path);
 
 	int i;
+
```
Comment (rpodgorny) [inline src/fuse_ops.c:406]: plase don't add whitespace commits

File: src/fuse_ops.c [hunk 5]
```diff
@@ -256,6 +256,10 @@ static void *unionfs_init(struct fuse_conn_info *conn, struct fuse_config *cfg)
 			exit(1);
 		}
 	}
+	
+	if (uopt.direct_io)
```
Comment (rpodgorny) [inline src/fuse_ops.c:260]: plase make this single line ((( if (uopt.direct_io) cfg->direct_io = 1; ))) or keep the {}. also please add spaces around =.

File: src/fuse_ops.c [hunk 6]
```diff
@@ -109,7 +109,7 @@ static int unionfs_create(const char *path, mode_t mode, struct fuse_file_info *
 
 	int i = find_rw_branch_cutlast(path);
 	if (i == -1) RETURN(-errno);
-
+
```
Comment (rpodgorny) [inline src/fuse_ops.c:112]: please remove this whitespace change

**Review by rpodgorny — COMMENTED** [review]
reviewed, thanks! ...please see the comments - it's basically just aesthetic stuff. ;-)

**Comment by rpodgorny** [comment]
...also, please don't hesitate to add yourself to CREDITS file. ;-)

File: src/opts.c (hunk 1 above)
Comment (DUOLabs333) [reply src/opts.c:294]: Before committing, do you know why some programs have trouble if the direct_io option is enabled in a chroot? Bindfs (which I similarly patched) doesn't have this issue.

**Comment by DUOLabs333** [comment]
Before merging, do you know why some programs have trouble if the direct_io option is enabled in a chroot? Bindfs (which I similarly patched) doesn't have this issue.

**Comment by DUOLabs333** [comment]
Also, fixed

**Comment by aakefbs** [comment]
What do you mean with trouble in chroots? Direct io is handled in the kernel - should not matter if chroot or not.

**Comment by DUOLabs333** [comment]
For example, `xbps` complains that `ERROR: cannot access to pkgdb: No such device`, but if I disable `direct_io`, it works.

**Comment by aakefbs** [comment]
Is there a chance to strace it? kernel fuse does not allow MAP_SHARED if direct-io is enabled. See the linux fuse_file_mmap() function. chroot or not does not matter for that, though.

**Comment by DUOLabs333** [comment]
What should I be looking for in the strace?

**Comment by aakefbs** [comment]
Try to find out what exactly fails.

**Comment by DUOLabs333** [comment]
Ok -- this is made harder by the fact that Void Linux doesn't have too much documentation about this.

**Comment by DUOLabs333** [comment]
While I'll keep looking, maybe someone else sees something I don't: https://pastebin.com/NHq8fi48

**Comment by aakefbs** [comment]
This is a fuse log. I meant `strace -f xbps`. From the fuse log it is impossible to see what xbfs is failing at.

**Comment by DUOLabs333** [comment]
That makes far more sense --- @aakefbs Looking at the logs, you're right: it has something to do with mmap `mmap(NULL, 241664, PROT_READ, MAP_SHARED, 3, 0) = -1 ENODEV (No such device)`

**Comment by aakefbs** [comment]
Ok, I have this on my fuse todo list, but I won't make any promises. In my opinion a libfuse 'relaxed_directio' option is missing, which would set some like FOPEN_DIRECT_IO_RELAXED and fuse_file_mmap() should then not fail the request. Right now I'm on fuse performance (kernel and libfuse). I will see if I get to the mmap topic any time soon. 
In anyway, this is not a unionfs-fuse issue.

**Comment by DUOLabs333** [comment]
> FOPEN_DIRECT_IO_RELAXED 
So you're saying that opening files with this flag will cause the operation to pass through. 

Also, you work on libfuse? That's cool.

**Comment by aakefbs** [comment]
See for example this: https://github.com/libfuse/libfuse/pull/652 and this https://www.spinics.net/lists/linux-fsdevel/msg214814.html. I'm just in the middle of the waitq kernel patches...

**Comment by DUOLabs333** [comment]
I'm mostly using this as a way to access `/proc`, as fuse would just report an empty file since the size is reported to be 0. Is there any other way of accessing procfs?

**Comment by aakefbs** [comment]
Use schroot and let it mount proc and sys into your chroot?

**Comment by DUOLabs333** [comment]
Yeah, I just figured out that I can just mount proc directly into the mountpoint instead of through unionfs.

**Comment by DUOLabs333** [comment]
I guess if there's no other complaints, this can get merged in, right?

**Comment by DUOLabs333** [comment]
> Use schroot and let it mount proc and sys into your chroot?

The problem is that bind mounts appear to `rsync` as on the same filesystem, messing up backups.

**Comment by DUOLabs333** [comment]
It seems that simply removing the `FUSE_DIRECT_IO` check in `fuse_file_mmap` might do the trick, but I'm not sure how much breaks as a result.

**Comment by DUOLabs333** [comment]
In the Linux source code, it says that if direct_io is set, return ENODEV. So, this a kernel problem, not a userspace problem. Are you going to merge a change that disables this behavior with a flag?

**Comment by aakefbs** [comment]
Yes that is the plan. But I'm not fuse upstream, just got a bit time at work to contribute a bit.  I know about this mmap issue for some time, but it is not a priority for work, at lot not for now.  We also have quite some issues to get our other fuse patches through - I cannot promise you that a fix for that will be merged any time soon.

I'm not sure what you are actually doing,but so far it sounds like your issue could be solved by other measures.

**Comment by DUOLabs333** [comment]
My main problem is that I want to be able to bind mount in such a way that it is recognized as a separate filesystem, so it won't be duplicated by rsync. FUSE does this, but honestly, it doesn't have to be userspace.

**Comment by aakefbs** [comment]
I guess I'm totally lost of what you are doing. Example from my system

```
bernd@t1700bs ~>mount
sysfs on /sys type sysfs (rw,relatime)
proc on /proc type proc (rw,relatime)
udev on /dev type devtmpfs (rw,nosuid,relatime,size=8117944k,nr_inodes=2029486,mode=755,inode64)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /run type tmpfs (rw,nosuid,nodev,noexec,relatime,size=1629056k,mode=755,inode64)
/dev/mapper/ubuntu--vg-root on / type ext4 (rw,relatime,stripe=32744)
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,inode64)
tmpfs on /run/lock type tmpfs (rw,nosuid,nodev,noexec,relatime,size=5120k,inode64)
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot)
pstore on /sys/fs/pstore type pstore (rw,nosuid,nodev,noexec,relatime)
none on /sys/fs/bpf type bpf (rw,nosuid,nodev,noexec,relatime,mode=700)
systemd-1 on /proc/sys/fs/binfmt_misc type autofs (rw,relatime,fd=29,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=20132)
mqueue on /dev/mqueue type mqueue (rw,nosuid,nodev,noexec,relatime)
hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime,pagesize=2M)
debugfs on /sys/kernel/debug type debugfs (rw,nosuid,nodev,noexec,relatime)
configfs on /sys/kernel/config type configfs (rw,nosuid,nodev,noexec,relatime)
fusectl on /sys/fs/fuse/connections type fusectl (rw,nosuid,nodev,noexec,relatime)
tracefs on /sys/kernel/tracing type tracefs (rw,nosuid,nodev,noexec,relatime)
sunrpc on /run/rpc_pipefs type rpc_pipefs (rw,relatime)
nfsd on /proc/fs/nfsd type nfsd (rw,relatime)
tmpfs on /run/qemu type tmpfs (rw,nosuid,nodev,relatime,mode=755,inode64)
tmpfs on /tmp type tmpfs (rw,relatime,inode64)
/var/lib/snapd/snaps/gnome-3-28-1804_145.snap on /snap/gnome-3-28-1804/145 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/gnome-3-28-1804_161.snap on /snap/gnome-3-28-1804/161 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/gtk-common-themes_1519.snap on /snap/gtk-common-themes/1519 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/teams_6.snap on /snap/teams/6 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/bare_5.snap on /snap/bare/5 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/gtk-common-themes_1515.snap on /snap/gtk-common-themes/1515 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/core18_2284.snap on /snap/core18/2284 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/gnome-3-38-2004_87.snap on /snap/gnome-3-38-2004/87 type squashfs (ro,nodev,relatime,x-gdu.hide)
/dev/mapper/ubuntu--vg-home2 on /home type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/dev/mapper/ubuntu--vg-home2 on /exports/home type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/dev/mapper/ubuntu--vg-home2 on /home/bschubert type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/dev/mapper/ubuntu--vg-home2 on /exports/home/bschubert type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/dev/mapper/ubuntu--vg-scratch on /scratch type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/dev/mapper/ubuntu--vg-scratch on /exports/scratch type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/dev/mapper/ubuntu--vg-kvm on /kvm type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/dev/mapper/ubuntu--vg-chroot on /chroots type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/dev/sda6 on /boot type ext4 (rw,relatime)
binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,nosuid,nodev,noexec,relatime)
tmpfs on /run/user/4003 type tmpfs (rw,nosuid,nodev,relatime,size=1629052k,nr_inodes=407263,mode=700,uid=4003,gid=4003,inode64)
gvfsd-fuse on /run/user/4003/gvfs type fuse.gvfsd-fuse (rw,nosuid,nodev,relatime,user_id=4003,group_id=4003)
tmpfs on /run/snapd/ns type tmpfs (rw,nosuid,nodev,noexec,relatime,size=1629056k,mode=755,inode64)
/var/lib/snapd/snaps/gnome-3-38-2004_99.snap on /snap/gnome-3-38-2004/99 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/snapd_14978.snap on /snap/snapd/14978 type squashfs (ro,nodev,relatime,x-gdu.hide)
portal on /run/user/4003/doc type fuse.portal (rw,nosuid,nodev,relatime,user_id=4003,group_id=4003)
/dev/mapper/ubuntu--vg-home on /media/bernd/home type ext4 (rw,nosuid,nodev,relatime,errors=remount-ro,uhelper=udisks2)
/var/lib/snapd/snaps/core20_1376.snap on /snap/core20/1376 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/snapd_15177.snap on /snap/snapd/15177 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/core18_2344.snap on /snap/core18/2344 type squashfs (ro,nodev,relatime,x-gdu.hide)
nsfs on /run/snapd/ns/teams.mnt type nsfs (rw)
tmpfs on /run/user/139 type tmpfs (rw,nosuid,nodev,relatime,size=1629052k,nr_inodes=407263,mode=700,uid=139,gid=148,inode64)
gvfsd-fuse on /run/user/139/gvfs type fuse.gvfsd-fuse (rw,nosuid,nodev,relatime,user_id=139,group_id=148)
/var/lib/snapd/snaps/chromium_1945.snap on /snap/chromium/1945 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/core20_1405.snap on /snap/core20/1405 type squashfs (ro,nodev,relatime,x-gdu.hide)
/var/lib/snapd/snaps/chromium_1951.snap on /snap/chromium/1951 type squashfs (ro,nodev,relatime,x-gdu.hide)
nsfs on /run/snapd/ns/chromium.mnt type nsfs (rw)
```

```
bernd@t1700bs ~>schroot -c debian-stable
bernd@debian@t1700bs ~>mount
/dev/mapper/ubuntu--vg-chroot on / type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
proc on /proc type proc (rw,relatime)
sysfs on /sys type sysfs (rw,relatime)
udev on /dev type devtmpfs (rw,nosuid,relatime,size=8117944k,nr_inodes=2029486,mode=755,inode64)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,inode64)
/dev/mapper/ubuntu--vg-home2 on /home type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
tmpfs on /tmp type tmpfs (rw,relatime,inode64)
/dev/mapper/ubuntu--vg-home2 on /home/bschubert type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/dev/mapper/ubuntu--vg-root on /var/yp/binding type ext4 (rw,relatime,stripe=32744)
tmpfs on /run type tmpfs (rw,nosuid,nodev,noexec,relatime,size=1629056k,mode=755,inode64)
tmpfs on /run/user/4003 type tmpfs (rw,nosuid,nodev,relatime,size=1629052k,nr_inodes=407263,mode=700,uid=4003,gid=4003,inode64)
```
So you can see that 'mount' actually shows the right entries in the chroot, as /etc/mtab is a link pointing to /proc/mounts.


Now, I don't get at all what is the issue with rsync here, even if mount/mtab would not show right entries, you have all the options to just filter out /proc, etc?

Anyway, this is totally out of topic for unionfs-fuse, but more generic linux questions...

**Comment by DUOLabs333** [comment]
> Anyway, this is totally out of topic for unionfs-fuse, but more generic linux questions...

Agreed, but I meant that bind mounts are still seen as on the same filesystem (like an upgraded version of symlinks). I could just manually exclude the bind mounts, though I would probably need to make a lot of them.

**Comment by rpodgorny** [comment]
@DUOLabs333 thank you for your corrections. i plan to review this and merge in the upcoming days (if there are no further objections)...

...btw, what would be the simplest way for me to test if it actually works? where exactly do you see the difference in your use case?

**Comment by DUOLabs333** [comment]
Mount proc with and without the option on diff. Try to access diff/self/stat. Only the mount with direct_io should return anything.

**Comment by aakefbs** [comment]
Well, I would classify that reproducer as fuse kernel bug. Reading /proc/schedstat over unionfs-fuse without -odirect_io (ubuntu version, package not updated yet (the update to the current master branch is hanging in debian-mentors)):

```
open flags: 0x8000 /schedstat
   open[7] flags: 0x8000 /schedstat
   unique: 14, success, outsize: 32
unique: 16, opcode: READ (15), nodeid: 2, insize: 80, pid: 14039
read[7] 4096 bytes from 0 flags: 0x8000
   read[7] 904 bytes from 0
   unique: 16, success, outsize: 920
```

So libfuse returns 4096 bytes, but fuse kernel then ignores that result. The reproducer *currently* works, but I don't think that will be a permanent test - fuse kernel really needs to be fixed...

**Comment by DUOLabs333** [comment]
You say "fuse kernel" --- is there a userspace version of FUSE?

**Comment by DUOLabs333** [comment]
Also, yes, but that's the reason I made the change, so I don't know where else it would be used.

**Comment by rpodgorny** [comment]
> Mount proc with and without the option on diff. Try to access diff/self/stat. Only the mount with direct_io should return anything.

this is a good candidate for the tests i've mentioned in https://github.com/rpodgorny/unionfs-fuse/issues/116

**Pull request closed without merging**